### DISCUSSION TITLE
multiple canvas instances

### DISCUSSION THREAD

**Discussion by mohammad-torabi-84** [moononournation/Arduino_GFX#346] [Q&A]
Hi,

for my project I need two instances of Arduino_Canvas. however, when I added the second canvas, the board keeps rebooting and it doesn't matter the order of the canvas or the size of them.

the board that I use is ESP32-8048S070, the dimension of the canvas 244*65 and 500*240.

the error messages in the Serial output is

20:04:48.900 -> �ESP-ROM:esp32s3-20210327
20:04:48.900 -> Build:Mar 27 2021
20:04:48.900 -> rst:0xc (RTC_SW_CPU_RST),boot:0x2a (SPI_FAST_FLASH_BOOT)
20:04:48.900 -> Saved PC:0x420961e6
20:04:48.900 -> SPIWP:0xee
20:04:48.900 -> mode:DIO, clock div:1
20:04:48.900 -> load:0x3fce3808,len:0x44c
20:04:48.900 -> load:0x403c9700,len:0xbe4
20:04:48.900 -> load:0x403cc700,len:0x2a68
20:04:48.900 -> entry 0x403c98d4
20:04:49.118 -> E (37) lcd_panel.rgb: esp_lcd_new_rgb_panel(128): no free rgb panel slot
20:04:49.118 -> ESP_ERROR_CHECK failed: esp_err_t 0x105 (ESP_ERR_NOT_FOUND) at 0x4037d9b8
20:04:49.164 -> file: "c:\Users\16477\OneDrive\Documents\HomeAutomation\Arduino\libraries\GFX_Library_for_Arduino\src\databus\Arduino_ESP32RGBPanel.cpp" line 118

is there any setting that I missed?

Thanks,

**Comment by moononournation** [comment]
I don't think you need to do it. What you actually want to do?

**Reply by mohammad-torabi-84** [reply]
I want to use one of the canvas to show a digital clock and the other one to show a timeline of events that changes every minutes and should be swiped.

**Reply by moononournation** [reply]
You should consider using LVGL

**Comment by rdekens** [comment]
similar here - appears the canvas class is tied to a display - I was wanting to use multiple canvas(s) a sub displays that I could draw and push - like sprites. Canvas class doesnt seem to support it.
I was trying with sprites - but the example  indexed sprite is extremely limited in that I can only import a bitmap to it - not perform draw operations on is like  I can with a canvas. 
the ability to decouple a canvas from a display  - after all it's a frame buffer, and draw to it using all the usual methods, and then be able to clip and push it to any display would be more flexible, rather than having to bring in the overhead of lvgl.

Ray

**Reply by moononournation** [reply]
you may check sprite_gif example for reference

**Reply by rdekens** [reply]
Sprite_gif ususes the indexed_sprite - and whas an extremely limitted set of operations and does not inherit anything  from elsewhere - it is a specialised case and not very useful.

I wish to use the draw functions on subdisplays - eg a canvas  that has all the associated graphics functions, sort of like mini windows - that can then be pushed / flushed to the a main canvas or display when desired. 

for example - I have code that build a display in a 200x200 window ( for an EPD) that i wish to reuse.  Build the display in a 200x200 canvas, and then use this on the main display (480 x 480) at a certain location, along with some other canvas(s) at differnt locations.

This would allow the code base to remain much the same between the EPD version and the T-RGB version - the EPD uses a stepper motor and real physical needle for the gauge , whereas the T-RGB would allow for a sprite or canvas needle pointer.

The ability to decouple the canvas and treat it as a sub display would I think provide a lot more flexibilty for your users. The transparency function of a sprite added in would make it perfect - 

Ray

**Reply by moononournation** [reply]
sprite_gif is designed can run without PSRAM, you can use 16-bit color space instead if you want better image quality.

**Reply by rdekens** [reply]
yes - but I cant use the draw functions on the indexed sprite - whereas they are all available on a canvas - 
relay there is no point in a canvas  other than as a framebuffer if I can only have one .

**Reply by moononournation** [reply]
You are freely change the example code as you wish ;)

**Reply by rdekens** [reply]
think Ive got it - 
in response tot he original topic - you can define multiple Canvases  - but you need to canvas->begin(GFX_SKIP_OUTPUT_BEGIN))
so that it does not look to initialise the output device but only define the canvas buffer -

you can create them and set the output to be another canvas in fact - I have a master canvas and a number of sub "window" canvases which are created with the relevant offset locations in the main canvas. ( like sprites)

In response the the "sprite issue" - I've then gone on to define  a -> flushWithTrans() to allow the transparent color to be set for the flush so that the canvasses can be stacked up to create the final image in the master canvas ( again in a sprite fashion).

Ray
